common/lockfree_queue.h

```diff
@@ -493,35 +486,70 @@ class LockfreeSPSCRingQueue : public LockfreeRingQueueBase<T, N> {
     }
 
     size_t push_batch(const T* x, size_t n) {
+        return produce_push_batch(n, [&](T* p1, size_t n1, T* p2, size_t n2) {
+            if (n1) { memcpy(p1, x, n1 * sizeof(T)); x += n1; }
+            if (n2) { memcpy(p2, x, n2 * sizeof(T)); }
+        });
+    }
+
+    // Producer must be callable as produce(T* offset1, size_t n, T* offset2, size_t n),
+    // and produce (write) data directly to the provided buffer.
+    template<typename Producer>
+    size_t produce_push_batch(size_t n, Producer&& produce) {
         auto t = tail.load(std::memory_order_relaxed);
         n = std::min(
             n, Base::capacity - (t - head.load(std::memory_order_acquire)));
         if (n == 0) return 0;
         auto first_idx = idx(t);
         auto part_length = Base::capacity - first_idx;
         if (likely(part_length >= n)) {
-            memcpy(&slots[first_idx], x, sizeof(T) * n);
+            produce(&slots[first_idx], n, nullptr, 0);
         } else {
-            if (likely(part_length))
-                memcpy(&slots[first_idx], x, sizeof(T) * (part_length));
-            memcpy(&slots[0], x + part_length, sizeof(T) * (n - part_length));
+            produce(&slots[first_idx], part_length,
+                    &slots[0], n - part_length);
+        }
+        tail.store(t + n, std::memory_order_release);
+        return n;
+    }
+
+    template<typename Producer>
+    size_t produce_push_batch_fully(size_t n, Producer&& produce) {
+        auto t = tail.load(std::memory_order_relaxed);
+        if (n > Base::capacity - (t - head.load(std::memory_order_acquire))) return 0;
+        auto first_idx = idx(t);
+        auto part_length = Base::capacity - first_idx;
+        if (likely(part_length >= n)) {
+            produce(&slots[first_idx], n, nullptr, 0);
```

Comment:
might be better if using `iovec` as callback parameters instead of two pointers and length.